internal/cmd/agent/deployer/driftdetect/driftdetect.go

```diff
@@ -1,57 +1,147 @@
-// AllResources returns the resources that are deployed by the bundle deployment,
-// according to the helm release history. It adds to be deleted resources to
-// the list, by comparing the desired state to the actual state with apply.
-func (m *DriftDetect) AllResources(bd *fleet.BundleDeployment) (*helmdeployer.Resources, error) {
-	resources, err := m.deployer.Resources(bd.Name, bd.Status.Release)
+func (d *DriftDetect) Clear(bdKey string) error {
+	return d.trigger.Clear(bdKey)
+}
+
+func (d *DriftDetect) Refresh(logger logr.Logger, bdKey string, bd *fleet.BundleDeployment, resources *helmdeployer.Resources) (*fleet.BundleDeployment, error) {
+	logger = logger.WithName("DriftDetect")
+	if bd == nil {
+		return bd, d.trigger.Clear(bdKey)
+	}
+
+	logger.V(1).Info("Refreshing drift detection")
+
+	resources, err := d.allResources(bd, resources)
+	if err != nil {
+		return bd, err
+	}
+
+	if resources == nil {
+		return bd, nil
+	}
+
+	logger.V(1).Info("Adding OnChange for bundledeployment's resource list")
+	logger = logger.WithValues("key", bdKey, "initial resource version", bd.ResourceVersion)
+
+	handler := func(key string) {
+		handleID := rand.Int() % 10000 // nolint:gosec // Non-crypto use
+		logger := logger.WithValues("handleID", handleID, "triggered by", key)
+		err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+			// Can't enqueue directly, update bundledeployment instead
+			return d.requeueBD(logger, handleID, bd.Namespace, bd.Name)
+		})
+		if err != nil {
+			logger.Error(err, "Failed to trigger bundledeployment", "error", err)
+			return
+		}
+
+	}
+	return bd, d.trigger.OnChange(bdKey, resources.DefaultNamespace, handler, resources.Objects...)
+}
+
+func (d *DriftDetect) requeueBD(logger logr.Logger, handleID int, namespace string, name string) error {
+	bd := &fleet.BundleDeployment{}
+
+	err := d.upstreamReader.Get(context.Background(), client.ObjectKey{Name: name, Namespace: namespace}, bd)
+	if apierrors.IsNotFound(err) {
+		logger.Info("Bundledeployment is not found, can't trigger refresh")
+		return nil
+	}
 	if err != nil {
-		return nil, nil
+		logger.Error(err, "Failed to get bundledeployment, can't trigger refresh")
+		return nil
 	}
 
+	logger = logger.WithValues("resource version", bd.ResourceVersion)
+	logger.V(1).Info("Going to update bundledeployment to trigger re-sync")
+
+	// This mechanism of triggering requeues for changes is not ideal.
+	// It's a workaround since we can't enqueue directly from the trigger
+	// mini controller. Triggering via a status update is expensive.
+	// It's hard to compute a stable hash to make this idempotent, because
+	// the hash would need to be computed over the whole change. We can't
+	// just use the resource version of the bundle deployment. We would
+	// need to look at the deployed resources and compute a hash over them.
+	// However this status update happens for every changed resource, maybe
+	// multiple times per resource. It will also trigger on a resync.
```

Comment:
Happy to talk or brainstorm about this, but could it be an option to order all resources by `kind` + `key`, hash their resource versions and compare that against `SyncGeneration`? This should avoid unnecessary re-deployments, for example on fleet-agent restarts?

Also, we should consider using `Generation` instead of the `ResourceVersion`. So status updates are omitted and don't requeue bundle deployments?

Reply:
Yes let's talk. I'm currently assuming it's cheaper to trigger than to compute a proper hash.

Reply:
yes, let's talk about this

Reply:
> Yes let's talk. I'm currently assuming it's cheaper to trigger than to compute a proper hash.

I'm not sure it's cheaper, especially since BundleDeployments updates may trigger as well updates on other resources on the upstream cluster (e.g. updating status of Bundles, GitRepos or Clusters resources). Also, beware of the multiplying factor in the case of many downstream clusters.